tests/integration/test_s3.py

```diff
@@ -1033,5 +1034,54 @@ def setUp(self):
         self.client = self.create_client()
 
 
-if __name__ == '__main__':
-    unittest.main()
+class TestRegionRedirect(BaseS3ClientTest):
+    def setUp(self):
+        super(TestRegionRedirect, self).setUp()
+        self.bucket_region = 'eu-central-1'
+        self.client_region = 'us-west-2'
+
+        self.client = self.session.create_client(
+            's3', region_name=self.client_region,
+            config=Config(signature_version='s3v4'))
+
+    def create_bucket(self, region_name, bucket_name=None):
+        saved_client = self.client
+        if region_name != self.client.meta.region_name:
+            self.client = self.session.create_client(
+                's3', region_name=region_name,
+                config=Config(signature_version='s3v4'))
+        bucket = super(TestRegionRedirect, self).create_bucket(
+            region_name, bucket_name)
+        self.client = saved_client
+        return bucket
+
+    def test_region_redirects(self):
+        self.bucket = self.create_bucket(self.bucket_region)
+        try:
+            response = self.client.list_objects(Bucket=self.bucket)
+            self.assertEqual(
+                response['ResponseMetadata']['HTTPStatusCode'], 200)
+        except ClientError:
```

Comment:
Should this check be more precise?  Seems like we should ensure we don't get the bucket is in wrong region code and allow any other errors to propagate.